client/util/Breadcrumbs.tsx

```diff
@@ -0,0 +1,87 @@
+export interface TreeNode {
+    pid: string;
+    title: string;
+    parents: Array<TreeNode>;
+}
+
+export interface TreeData {
+    topNodes: Array<string>;
+    records: Record<string, TreeNode>;
+    childLookups: Record<string, Array<string>>;
+}
+
+interface BreadcrumbTrail {
+    pid: string;
+    path: Array<TreeNode>;
+}
+
+/**
+ * Analyze the raw breadcrumb data, identifying top-level nodes (i.e. nodes with
+ * no parents) and creating lookup tables for children. This makes it easier to
+ * render breadcrumbs from left to right and to find all relevant trails.
+ */
+export function processBreadcrumbData(data: TreeNode): TreeData {
+    const queue: Array<TreeNode> = [data];
+    const topNodes: Array<string> = [];
+    const childLookups: Record<string, Array<string>> = {};
+    const records: Record<string, TreeNode> = {};
+    while (queue.length > 0) {
+        const current = queue.shift();
+        if (!current) {
+            break;
+        }
+        if (current.parents.length === 0) {
+            topNodes.push(current.pid);
+        }
+        records[current.pid] = { pid: current.pid, title: current.title, parents: [] };
+        current.parents.forEach((parent) => {
+            queue.push(parent);
+            if (typeof childLookups[parent.pid] === "undefined") {
+                childLookups[parent.pid] = [current.pid];
+            } else {
+                childLookups[parent.pid].push(current.pid);
+            }
+        });
+    }
+    return {
+        topNodes: Array.from(new Set(topNodes)), // deduplicate
```

Comment:
You can maybe instantiate topNodes into Set first.

Reply:
So you're just suggesting that I define topNodes as a Set instead of an Array so there's no need to convert back and forth? If that's the case, it makes sense to me... but again, I will probably address that as a separate PR since it doesn't directly relate to the work in this one.

Reply:
Fixed in #165.